[SPARK-10810] [SPARK-10902] [SQL] Improve session management in SQL

davies opened the pull request:
This PR improve the sessions management by replacing the thread-local based to one SQLContext per session approach, introduce separated temporary tables and UDFs/UDAFs for each session.

A new session of SQLContext could be created by:

1) create an new SQLContext
2) call newSession() on existing SQLContext

For HiveContext, in order to reduce the cost for each session, the classloader and Hive client are shared across multiple sessions (created by newSession).

CacheManager is also shared by multiple sessions, so cache a table multiple times in different sessions will not cause multiple copies of in-memory cache.

Added jars are still shared by all the sessions, because SparkContext does not support sessions.

cc @marmbrus @yhuai @rxin

Commit:
sessions for SQLContext

Commit:
improve executionHive

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42982 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42982/consoleFull) for   PR 8909 at commit [`bc9f064`](https://github.com/apache/spark/commit/bc9f06447482dac5cfc61fd9d8c21bacb2300431).

SparkQA commented:
[Test build #42982 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42982/console) for   PR 8909 at commit [`bc9f064`](https://github.com/apache/spark/commit/bc9f06447482dac5cfc61fd9d8c21bacb2300431).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42982/
Test FAILed.

Commit:
fix scala style and current_database

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42988 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42988/consoleFull) for   PR 8909 at commit [`0cf26a5`](https://github.com/apache/spark/commit/0cf26a57c844bd7379e07f518555418b0b3ecefb).

SparkQA commented:
[Test build #42988 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42988/console) for   PR 8909 at commit [`0cf26a5`](https://github.com/apache/spark/commit/0cf26a57c844bd7379e07f518555418b0b3ecefb).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42988/
Test FAILed.

SparkQA commented:
[Test build #1814 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1814/consoleFull) for   PR 8909 at commit [`0cf26a5`](https://github.com/apache/spark/commit/0cf26a57c844bd7379e07f518555418b0b3ecefb).

AmplabJenkins commented:
Merged build triggered.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:
```diff
@@ -68,6 +70,14 @@ class SimpleFunctionRegistry extends FunctionRegistry {
     functionBuilders.get(name).map(_._1)
   }
+
+  override def copy(): SimpleFunctionRegistry = {
```
should this be synchronized?

davies replied:
SimpleFunctionRegistry is not designed to be used by multiple threads (read/write in the same time), I think it's fine now, because the builtin is constructed in the begging.

rxin replied:
OK this is technically not part of this, but since you are modifying it ... shouldn't all of these catalogs be thread safe? Since they can be called by multiple threads.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42991 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42991/consoleFull) for   PR 8909 at commit [`3723768`](https://github.com/apache/spark/commit/372376862ccbf1877dbbb2a26761c76d4b3afdd5).

SparkQA commented:
[Test build #1814 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1814/console) for   PR 8909 at commit [`0cf26a5`](https://github.com/apache/spark/commit/0cf26a57c844bd7379e07f518555418b0b3ecefb).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #42991 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42991/console) for   PR 8909 at commit [`3723768`](https://github.com/apache/spark/commit/372376862ccbf1877dbbb2a26761c76d4b3afdd5).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42991/
Test FAILed.

Commit:
fix mima

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42997 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42997/consoleFull) for   PR 8909 at commit [`9c01c27`](https://github.com/apache/spark/commit/9c01c27f9a8b96a86f2260a2aeb8f942771b054c).

SparkQA commented:
[Test build #42997 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42997/console) for   PR 8909 at commit [`9c01c27`](https://github.com/apache/spark/commit/9c01c27f9a8b96a86f2260a2aeb8f942771b054c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42997/
Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into sessions

Conflicts:
	sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43135 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43135/consoleFull) for   PR 8909 at commit [`5160fea`](https://github.com/apache/spark/commit/5160fea483e7a3b224b20aef01cb371898c0d5cf).

SparkQA commented:
[Test build #43135 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43135/console) for   PR 8909 at commit [`5160fea`](https://github.com/apache/spark/commit/5160fea483e7a3b224b20aef01cb371898c0d5cf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class HyperLogLogPlusPlus(child: Expression, relativeSD: Double = 0.05)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43135/
Test FAILed.

Commit:
fix some tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43137 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43137/consoleFull) for   PR 8909 at commit [`e942e93`](https://github.com/apache/spark/commit/e942e9333274527665d9a72d4a9ceb14ae711854).

SparkQA commented:
[Test build #43137 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43137/console) for   PR 8909 at commit [`e942e93`](https://github.com/apache/spark/commit/e942e9333274527665d9a72d4a9ceb14ae711854).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class HyperLogLogPlusPlus(child: Expression, relativeSD: Double = 0.05)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43137/
Test FAILed.

Commit:
fix tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43141 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43141/consoleFull) for   PR 8909 at commit [`bcaddb3`](https://github.com/apache/spark/commit/bcaddb31cbeb10d8eb06a5397b5059417818b942).

SparkQA commented:
[Test build #43141 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43141/console) for   PR 8909 at commit [`bcaddb3`](https://github.com/apache/spark/commit/bcaddb31cbeb10d8eb06a5397b5059417818b942).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class HyperLogLogPlusPlus(child: Expression, relativeSD: Double = 0.05)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43141/
Test FAILed.

Commit:
refactor

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43154 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43154/consoleFull) for   PR 8909 at commit [`2e365ad`](https://github.com/apache/spark/commit/2e365ada232e42d88692c89634e5ed2ceb741beb).

SparkQA commented:
[Test build #43154 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43154/console) for   PR 8909 at commit [`2e365ad`](https://github.com/apache/spark/commit/2e365ada232e42d88692c89634e5ed2ceb741beb).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class HyperLogLogPlusPlus(child: Expression, relativeSD: Double = 0.05)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43154/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43155 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43155/consoleFull) for   PR 8909 at commit [`9fe30cf`](https://github.com/apache/spark/commit/9fe30cf6372fd5c78a6aae004e19bd038f1cc93f).

Commit:
add more tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43158 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43158/consoleFull) for   PR 8909 at commit [`69f87cb`](https://github.com/apache/spark/commit/69f87cb1bd7d243a7a2e8e6eccd45c25845efac1).

SparkQA commented:
[Test build #43158 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43158/console) for   PR 8909 at commit [`69f87cb`](https://github.com/apache/spark/commit/69f87cb1bd7d243a7a2e8e6eccd45c25845efac1).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  /** Run a function within Hive state (SessionState, HiveConf, Hive client and class loader) */`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43158/
Test FAILed.

SparkQA commented:
[Test build #43155 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43155/console) for   PR 8909 at commit [`9fe30cf`](https://github.com/apache/spark/commit/9fe30cf6372fd5c78a6aae004e19bd038f1cc93f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43155/
Test FAILed.

Commit:
fix test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43160 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43160/consoleFull) for   PR 8909 at commit [`d10a97a`](https://github.com/apache/spark/commit/d10a97a23608e2f27bbeb6760559972b26d64c56).

SparkQA commented:
[Test build #43160 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43160/console) for   PR 8909 at commit [`d10a97a`](https://github.com/apache/spark/commit/d10a97a23608e2f27bbeb6760559972b26d64c56).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  /** Run a function within Hive state (SessionState, HiveConf, Hive client and class loader) */`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43160/
Test FAILed.

Commit:
fix tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43162 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43162/consoleFull) for   PR 8909 at commit [`9b82c55`](https://github.com/apache/spark/commit/9b82c55090080ca34b2b9e271c08afdeb0dd25e6).

SparkQA commented:
[Test build #43162 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43162/console) for   PR 8909 at commit [`9b82c55`](https://github.com/apache/spark/commit/9b82c55090080ca34b2b9e271c08afdeb0dd25e6).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  /** Run a function within Hive state (SessionState, HiveConf, Hive client and class loader) */`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43162/
Test FAILed.

Commit:
fix mima

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43165 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43165/consoleFull) for   PR 8909 at commit [`72e137d`](https://github.com/apache/spark/commit/72e137d146db9e77efe0d64bfa5bb909d16654c2).

SparkQA commented:
[Test build #43165 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43165/console) for   PR 8909 at commit [`72e137d`](https://github.com/apache/spark/commit/72e137d146db9e77efe0d64bfa5bb909d16654c2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  /** Run a function within Hive state (SessionState, HiveConf, Hive client and class loader) */`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43165/
Test FAILed.

Commit:
bug fix

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43169 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43169/consoleFull) for   PR 8909 at commit [`3b48dd8`](https://github.com/apache/spark/commit/3b48dd8c9d3beb28c03dd6cdf6cf9e629a1645fd).

SparkQA commented:
[Test build #43169 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43169/console) for   PR 8909 at commit [`3b48dd8`](https://github.com/apache/spark/commit/3b48dd8c9d3beb28c03dd6cdf6cf9e629a1645fd).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  /** Run a function within Hive state (SessionState, HiveConf, Hive client and class loader) */`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43169/
Test PASSed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -300,21 +303,25 @@ class SQLContext(@transient val sparkContext: SparkContext)
    * @since 1.3.0
    */
-  def uncacheTable(tableName: String): Unit = cacheManager.uncacheTable(tableName)
+  def uncacheTable(tableName: String): Unit = cacheManager.uncacheQuery(table(tableName))
```
We can also add `cacheQuery` here so that we don't need to pass in a `SQLContext` every time we call `CacheManager.cacheQuery `

davies replied:
done

Commit:
fix current_database()

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43287 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43287/consoleFull) for   PR 8909 at commit [`ae0e876`](https://github.com/apache/spark/commit/ae0e87677a6c27532f071c4f064d42805029b53d).

SparkQA commented:
[Test build #43287 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43287/console) for   PR 8909 at commit [`ae0e876`](https://github.com/apache/spark/commit/ae0e87677a6c27532f071c4f064d42805029b53d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  /** Run a function within Hive state (SessionState, HiveConf, Hive client and class loader) */`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43287/
Test PASSed.

navis commented:
@davies I love the idea removing thread-local, which caused headache for me doing #7118. One thing I like to see also fixed in here is that jdbc meta calls like "connection.getMetaData" is not working currently because it's executed via exec-hive, which sees dummy-derby metastore.

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:
```diff
@@ -51,22 +51,42 @@ class SimpleFunctionRegistry extends FunctionRegistry {
+      name: String,
+      info: ExpressionInfo,
+      builder: FunctionBuilder): Unit = {
+    synchronized {
```
Style nit: in most places we put `synchronized` inline with the return type.

davies replied:
In copy(), `iterator.foreach` can be protected by the synchronization of map.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -64,18 +63,28 @@ import org.apache.spark.util.Utils
 
   /**
+   * Returns a SQLContext as new session, with separated SQL configurations, temporary tables,
+   * registered functions, but share the same SparkContext and CacheManager.
```
nit: "sharing"

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1208,36 +1204,72 @@ object SQLContext {
-   * Reference to the last created SQLContext.
+   * The active SQLContext for threads.
+   */
+  private val activeContexts: InheritableThreadLocal[SQLContext] =
```
Nit: I think this should be singular since there is only one active context per thread.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1208,36 +1204,72 @@ object SQLContext {
 
   /**
-   * Reference to the last created SQLContext.
+   * The active SQLContext for threads.
```
The active SQLContext for _the current thread_.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1208,36 +1204,72 @@ object SQLContext {
+   * Reference to the created SQLContext.
    */
-  @transient private val lastInstantiatedContext = new AtomicReference[SQLContext]()
+  @transient private val instantiatedContext = new AtomicReference[SQLContext]()
```
I think this makes more sense than "last" but we should update the objects doc if we are going to change the semantics.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1208,36 +1204,72 @@ object SQLContext {
     }
   }
 
+  /*
```
nit: indent and need two *

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala:
```diff
@@ -89,6 +80,7 @@ private[sql] class CacheManager(sqlContext: SQLContext) extends Logging {
    * the in-memory columnar representation of the underlying table is expensive.
    */
   private[sql] def cacheQuery(
+      sqlContext: SQLContext,
```
You can get the `sqlContext` from `query` right?

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1208,36 +1204,72 @@ object SQLContext {
   }
 
+  /*
+    * Set a SQLContext for current thread
```
"Changes the SQLContext that will be returned in this thread and its children when `SQLContext.getOrCreate()` is called.  This can be used to ensure that a given thread receives a SQLContext with an isolated session, instead of the global (first created) context."

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1208,36 +1204,72 @@ object SQLContext {
+  }
+
+  /**
+   * Clear the active SQLContext for current thread
```
"Clears the active SQLContext for current thread.  Subsequent calls to getOrCreate will return the first created context instead of a thread-local override."

marmbrus commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:
```diff
@@ -159,7 +159,7 @@ class CliSuite extends SparkFunSuite with BeforeAndAfter with Logging {
         -> "OK",
       "CACHE TABLE hive_test;"
-        -> "Time taken: ",
+        -> "",
```
What's the reason for these changes?

davies replied:
The output does not always have `Time taken`, test will fail.

marmbrus commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala:
```diff
@@ -289,16 +293,15 @@ class HiveThriftBinaryServerSuite extends HiveThriftJdbcTest {
         }
-        rs1.close()
 
+        // TODO need to figure out how to determine if the data loaded from cache
```
You can grab the queryExecution.executedPlan and look for `InMemoryColumnarTableScan`

davies replied:
This is JDBC client, so we can't access the internal of catalyst. Maybe we could use explain().

marmbrus commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala:
```diff
@@ -148,53 +150,88 @@ private[hive] class IsolatedClientLoader(
+    }
   }
+
+  private[hive] var cachedHive: Any = null
```
Can you document what this is used for.

marmbrus commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala:
```diff
@@ -1133,6 +1133,17 @@ class HiveQuerySuite extends HiveComparisonTest with BeforeAndAfter {
+    s2.sql("use b")
+
+    assert(sql("select current_database()").first() === Row("a"))
+    assert(s2.sql("select current_database()").first() === Row("b"))
```
Can we also add a test that tries querying from tables that only exist in one database or another to make sure that the catalogs resolution of "current database" is working.

davies replied:
That is already covered by HiveThriftServerSuite, could also addd that here.

marmbrus replied:
Yeah, the thrift server tests are so expensive / flakey that I'd like to have another cheaper unit test for all of this.

davies replied:
Added.

marmbrus commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala:
```diff
@@ -116,27 +116,19 @@ class TestHiveContext(sc: SparkContext) extends HiveContext(sc) {
-      override def dialect: String = super.getConf(SQLConf.DIALECT, "hiveql")
-      override def caseSensitiveAnalysis: Boolean = getConf(SQLConf.CASE_SENSITIVE, false)
+  protected[sql] override lazy val conf: SQLConf = new SQLConf {
+    // TODO as in unit test, conf.clear() probably be called, all of the value will be cleared.
```
existing: is this TODO relevant still?

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -64,18 +63,28 @@ import org.apache.spark.util.Utils
    * @return Spark SQL configuration
    */
-  protected[sql] def conf = currentSession().conf
+  protected[sql] lazy val conf = new SQLConf
```
Can we avoid of using `lazy val`?

davies replied:
`conf` will be override in HiveContext, but the constructor of SQLContext need to access that, so it can't be `val`.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -64,18 +63,28 @@ import org.apache.spark.util.Utils
 
   self =>
 
+  def this(sparkContext: SparkContext) = this(sparkContext, new CacheManager)
```
Let's add scaladoc to explain when it is recommended to use this constructor and when it recommended to use `newSession`.

marmbrus commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala:
```diff
@@ -410,7 +436,11 @@ class HiveContext(sc: SparkContext) extends SQLContext(sc) with Logging {
+
+  functionRegistry.registerFunction(
+    "current_database",
+    (expressions: Seq[Expression]) => new CurrentDatabase(this))
```
This is a cute way to tie this to the session.

yhuai replied:
Add a comment to explain why we need to do it at here?

marmbrus commented:
@davies, this is great.  Much cleaner than the old implementation!

marmbrus commented:
@navis, that does sound like something we should fix, but unless we are breaking that here (and I don't think we are).  I'd rather do this in a followup PR rather than block this large easily conflicted change.  Please correct me if I'm wrong though.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1208,36 +1204,72 @@ object SQLContext {
+    }
+  }
+
+  private[sql] def clearTheInstantiatedContext(sqlContext: SQLContext): Unit = {
```
Should we change the name of of `clearTheInstantiatedContext` to reflect its difference with `clearInstantiatedContext`?

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1208,36 +1204,72 @@ object SQLContext {
+  def clearActive(): Unit = {
+    activeContexts.remove()
+  }
+
```
Can we have a section in the scaladoc to give an overview of these newly added methods?

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala:
```diff
@@ -69,13 +84,24 @@ private[hive] class HiveQLDialect extends ParserDialect {
+    sc: SparkContext,
+    cacheManager: CacheManager,
+    @transient execHive: ClientWrapper,
+    @transient metaHive: ClientInterface) extends SQLContext(sc, cacheManager) with Logging {
```
Maybe add comments to explain what `execHive` and `metaHive` do?

davies replied:
This is not public API. Those two are used to create a new session with existing clients (forked from another one).

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala:
```diff
@@ -136,7 +137,8 @@ private[hive] class IsolatedClientLoader(
     name.startsWith("java.net") ||
-    sharedPrefixes.exists(name.startsWith)
+    sharedPrefixes.exists(name.startsWith) ||
+    name.startsWith("[")
```
Since it is not super obvious, can we add a comment to say that this is for arrays?

marmbrus replied:
Haha, yeah that took me a second.  Any idea why this is needed now? (it seems reasonable to add).

davies replied:
Just saw that many array types are loaded multiple times, it's not that necessary for this PR.

marmbrus replied:
Thanks for the explanation.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala:
```diff
@@ -148,53 +150,88 @@ private[hive] class IsolatedClientLoader(
-      if (loaded == null) doLoadClass(name, resolve) else loaded
-    }
+  private[hive] var classLoader: ClassLoader = if (isolationOn) {
+    new URLClassLoader(allJars, rootClassLoader) {
```
Since we are overriding more stuff, should we have a explicit class definition with docs about how it works and reasons that we need it?

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala:
```diff
@@ -148,53 +150,88 @@ private[hive] class IsolatedClientLoader(
-    }
+  private[hive] var classLoader: ClassLoader = if (isolationOn) {
+    new URLClassLoader(allJars, rootClassLoader) {
+      val cache = new java.util.concurrent.ConcurrentHashMap[String, Class[_]]
```
If we want to use it as a cache, should we use a data structure that defines the cache capacity?

davies replied:
will remove the cache in this PR.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala:
```diff
@@ -148,53 +150,88 @@ private[hive] class IsolatedClientLoader(
-  Thread.currentThread.setContextClassLoader(classLoader)
+  private[hive] def addJar(path: String): Unit = synchronized {
+    val jarURL = new java.io.File(path).toURI.toURL
+    classLoader = new java.net.URLClassLoader(Array(jarURL), classLoader)
```
While you are changing this part, can you add a TODO at here saying something like we should avoid of stacking classloaders (it will be good to use a single URLClassLoader and add jars to that).

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala:
```diff
@@ -264,12 +262,16 @@ private[hive] object HiveQl extends Logging {
+      val conf = new HiveConf(classOf[SessionState])
+      conf.setClassLoader(original)
+      ss = new SessionState(conf)
+      SessionState.start(ss)
```
`SessionState.start` will change the context classloader to the classloader associated with `conf`. Is it the expected behavior?

davies replied:
In the case of no SessionState for current thread, it will be NPE if we do not call `start()`.  Should this only used by tests?

yhuai replied:
Actually, I guess we are fine. When we create a `HiveConf`, its internal classloader will be `Thread.currentThread().getContextClassLoader()`. So, we will not change the context classloader. But, it is better to have explicitly call `conf.setClassLoader` to set `original` as the classloader for this conf to avoid any confusion. What do you think?

yhuai commented:
This is super nice! It will be great if we can manually test the following cases to make sure Hive UDFs and SerDes added through user jars are good:
* a spark-shell with jars loaded through `--jars` option.
* a spark-shell with jars loaded through `add jar` command.
* a beeline client with jars loaded through `add jar` command.
* a cli client with jars loaded through `add jar` command.
* a cli client with jars loaded through `--jars` option.

Regarding commands to test, we can try commands that involve UDFs (register a UDF and use it) and SerDes (define a table, write data to it and query it).

Also, can we put a diagram in the comment (maybe in a follow-up pr) to explain relationship among HiveContext, ClientWrapper, and IsolatedLoader?

Commit:
address comments

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1196,48 +1174,93 @@ class SQLContext(@transient val sparkContext: SparkContext)
    * This function can be used to create a singleton SQLContext object that can be shared across
    * the JVM.
+   *
+   * If there an active SQLContext for current thread, it will returned instead of the global one.
```
Nit: "If there **is** an active ..."

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

liancheng commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:
```diff
@@ -210,6 +190,9 @@ private[hive] class SparkExecuteStatementOperation(
     setState(OperationState.RUNNING)
+    val executionHiveClassLoader =
+      hiveContext.executionHive.state.getConf.getClassLoader
+    Thread.currentThread().setContextClassLoader(executionHiveClassLoader)
```
Would be nice to add a comment here to explain why we need to set the class loader here. (To make `add jar` work properly right?)

SparkQA commented:
[Test build #43350 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43350/consoleFull) for   PR 8909 at commit [`b70720f`](https://github.com/apache/spark/commit/b70720fa18fc6361d3ea9d394a32c2c603b2018f).

davies commented:
@yhuai `ADD JAR` with SERDE is already covered well by unit tests (including JDBC client and Cli and HiveContext). `--jar` is tested manually, work well.  Which jar could be used to test UDF?

yhuai commented:
@davies Thanks for checking! For UDF, we can use `hive-contrib-0.13.1.jar` (https://github.com/apache/spark/blob/master/sql/hive/src/test/resources/hive-contrib-0.13.1.jar) which contains a udaf called `org.apache.hadoop.hive.contrib.udaf.example.UDAFExampleMax`.

davies commented:
@yhuai UDF and UDTF are also covered by testes, also checked that manually, work fine.

yhuai commented:
Cool. Thanks!

Commit:
address comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

Commit:
add doc for newSession

SparkQA commented:
[Test build #43356 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43356/consoleFull) for   PR 8909 at commit [`e16737c`](https://github.com/apache/spark/commit/e16737c18f2662da1ddfd00feb9e5f433ec94184).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43357 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43357/consoleFull) for   PR 8909 at commit [`3aeae47`](https://github.com/apache/spark/commit/3aeae474fdada66b38b8ef81aa1523880f1b8a8a).

andrewor14 commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1196,49 +1174,101 @@ class SQLContext(@transient val sparkContext: SparkContext)
   // construction of the instance.
-  SQLContext.setLastInstantiatedContext(self)
+  sparkContext.addSparkListener(new SparkListener{
+    override def onApplicationEnd(applicationEnd: SparkListenerApplicationEnd) {
```
`Unit`, also space before `{` in previous line

davies replied:
@andrewor14 The `SparkListener` does not have `: Unit`.

marmbrus replied:
This is implicit `Unit`, but I think they are going to get rid of this in a future version of scala (and I think its against our style guide).

andrewor14 commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1196,49 +1174,101 @@ class SQLContext(@transient val sparkContext: SparkContext)
    * This function can be used to create a singleton SQLContext object that can be shared across
    * the JVM.
+   *
+   * If there is an active SQLContext for current thread, it will returned instead of the global
```
it will returned -> it will be returned

andrewor14 commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1196,49 +1174,101 @@ class SQLContext(@transient val sparkContext: SparkContext)
+   * The active SQLContext for the current thread.
    */
-  @transient private val lastInstantiatedContext = new AtomicReference[SQLContext]()
+  private val activeContext: InheritableThreadLocal[SQLContext] =
```
it's not clear how this is different from `instantiatedContext` from the name. Maybe this should be `threadLocalContext` or `threadActiveContext` or something

andrewor14 commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1196,49 +1174,101 @@ class SQLContext(@transient val sparkContext: SparkContext)
+  /**
+   * Reference to the created SQLContext.
+   */
+  @transient private val instantiatedContext = new AtomicReference[SQLContext]()
```
is it possible to have just one of these? It seems currently that we only use `activeContext` if the user explicitly called `setActive`. Can we just use the same `AtomicReference` for `setActive`?

andrewor14 replied:
if we absolutely have to keep both, then we should expand on the comments to explain why a thread local context is needed in addition to one that was initially instantiated.

SparkQA commented:
[Test build #43356 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43356/console) for   PR 8909 at commit [`e16737c`](https://github.com/apache/spark/commit/e16737c18f2662da1ddfd00feb9e5f433ec94184).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  /** Run a function within Hive state (SessionState, HiveConf, Hive client and class loader) */`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43356/
Test FAILed.

SparkQA commented:
[Test build #43357 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43357/console) for   PR 8909 at commit [`3aeae47`](https://github.com/apache/spark/commit/3aeae474fdada66b38b8ef81aa1523880f1b8a8a).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  /** Run a function within Hive state (SessionState, HiveConf, Hive client and class loader) */`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43357/
Test FAILed.

SparkQA commented:
[Test build #1856 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1856/consoleFull) for   PR 8909 at commit [`3aeae47`](https://github.com/apache/spark/commit/3aeae474fdada66b38b8ef81aa1523880f1b8a8a).

andrewor14 commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1196,49 +1174,101 @@ class SQLContext(@transient val sparkContext: SparkContext)
   }
 
-  private[sql] def clearLastInstantiatedContext(): Unit = {
+  private[sql] def clearInstantiatedContext(): Unit = {
```
this isn't used anymore except in tests is it? Do we still need it?

SparkQA commented:
[Test build #43350 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43350/console) for   PR 8909 at commit [`b70720f`](https://github.com/apache/spark/commit/b70720fa18fc6361d3ea9d394a32c2c603b2018f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  /** Run a function within Hive state (SessionState, HiveConf, Hive client and class loader) */`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43350/
Test FAILed.

SparkQA commented:
[Test build #1856 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1856/console) for   PR 8909 at commit [`3aeae47`](https://github.com/apache/spark/commit/3aeae474fdada66b38b8ef81aa1523880f1b8a8a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  /** Run a function within Hive state (SessionState, HiveConf, Hive client and class loader) */`

Commit:
fix tests and address comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43407 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43407/consoleFull) for   PR 8909 at commit [`19b91df`](https://github.com/apache/spark/commit/19b91dfc1c6ce16409411dbb54925dcebc39e608).

SparkQA commented:
[Test build #43407 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43407/console) for   PR 8909 at commit [`19b91df`](https://github.com/apache/spark/commit/19b91dfc1c6ce16409411dbb54925dcebc39e608).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  /** Run a function within Hive state (SessionState, HiveConf, Hive client and class loader) */`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43407/
Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into sessions

Conflicts:
	sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43428 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43428/consoleFull) for   PR 8909 at commit [`ab5e46c`](https://github.com/apache/spark/commit/ab5e46cc79c5d3fe4c1b0662ac58ed07f3b94f34).

davies commented:
@marmbrus @yhuai @andrewor14 I should have addressed all of your comments (except lack of more docs), could you take another pass? The failed test is not related to this.

SparkQA commented:
[Test build #43428 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43428/console) for   PR 8909 at commit [`ab5e46c`](https://github.com/apache/spark/commit/ab5e46cc79c5d3fe4c1b0662ac58ed07f3b94f34).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43428/
Test PASSed.

andrewor14 commented:
Non-hive parts LGTM!

davies commented:
Merged into master, new comments will be addressed in follow-up PR, thanks you all!

WangTaoTheTonic commented:
Nice work! As it is a blocker could we merge this into branch-1.5 too?

yhuai commented:
@WangTaoTheTonic Since this is a not a bug fix (instead, a major new feature), we will not merge it to branch-1.5. It will be released with Spark 1.6.0.

tedyu commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1196,46 +1174,90 @@ class SQLContext(@transient val sparkContext: SparkContext)
+  /**
+   * The active SQLContext for the current thread.
+   */
+  private val activeContext: InheritableThreadLocal[SQLContext] =
```
Should this be annotated @volatile ?

tedyu commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -1196,46 +1174,90 @@ class SQLContext(@transient val sparkContext: SparkContext)
+   * Reference to the created SQLContext.
    */
-  @transient private val lastInstantiatedContext = new AtomicReference[SQLContext]()
+  @transient private val instantiatedContext = new AtomicReference[SQLContext]()
```
Should this be annotated @volatile ?

rxin replied:
isn't AtomicReference itself just a wrapper around volatile field?

marmbrus replied:
Yeah agreed.  In fact @volatile should never be needed on a val.

tedyu replied:
Alright.
I was looking at getOrCreate() at line 1219
Seems the access to instantiatedContext is proper.

Neuw84 commented:
Since upgrading to Spark 1.6 I can´t query temp tables created from SparkStreaming using Apache Toad.

The relevant Spark Streaming part.
 
            window.foreachRDD((JavaRDD<Row> rdd, Time time) -> {
            if (!rdd.isEmpty()) {
                HiveContext sqlContext1 = JavaSQLContextSingleton.getInstance(rdd.context());
                if (sqlContext1.isCached("issues_recent")) {
                    sqlContext1.uncacheTable("issues_recent");
                }
                DataFrame recentDataFrame = sqlContext1.createDataFrame(rdd, schema).coalesce(20);
                recentDataFrame.registerTempTable("issues_recent");
                recentDataFrame.cache();             
            }

The SQLContext creation part
class JavaSQLContextSingleton {

    static private transient HiveContext instance = null;

    static public HiveContext getInstance(SparkContext sparkContext) {
        if (instance == null) {
            instance = new HiveContext(sparkContext);
        }
        return instance;
    }
}

Without using spark streaming doesn´t help with the problem (loading parque file a registering a temporary table).

I think that this "issue" may be related to this PR (maybe I am wrong and this is not the correct place to put this).

Any hints?

davies commented:
@Neuw84 Since 1.6, each SQLContext will have separated session, you should share the same SQLContext in this case (using SQLContext.getOrCreate()).

deenar commented:
@Neuw84, i presume you are connecting TOAD to Spark SQL via the thrift server. You might need to set  spark.sql.hive.thriftServer.singleSession=true to preserve the pre 1.6 behaviour

Neuw84 commented:
@deenar, I saw it after many hours reading the code in the web docs .  Although, I think that the HiveSparkContext should implement the same logic as the SparkContext where you can get the same Session programmatically. Thanks by the way

chutium commented:
anyone tried use --hivevar or --hiveconf with beeline after this change? variables from JDBC client is no more available in the hive session.
refer to  https://issues.apache.org/jira/browse/SPARK-13983

is there any test case for --hivevar and --hiveconf in test suites?

JoshRosen commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala:
```diff
@@ -598,6 +631,14 @@ class HiveContext(sc: SparkContext) extends SQLContext(sc) with Logging {
+    // Add jar to Hive and classloader
+    executionHive.addJar(path)
+    metadataHive.addJar(path)
+    Thread.currentThread().setContextClassLoader(executionHive.clientLoader.classLoader)
```
@davies @yhuai Do we need to restore the original classloader after calling `super.addJar()` or is it intentional that calling `HiveContext.addJar` will permanently mutate the calling thread's context classloader?
